File: python/2/main.py
from collections import Counter


def parse_line(line):
    return line.strip().replace("-", " ").replace(":", "").split(" ")


def part1(rows):
    
    count = 0
    
    for line in rows:
        rule_min, rule_max, letter, password = parse_line(line)
        
        letter_counter = Counter(password)

        # Check if the string is within the specification according to the policy
        if int(rule_min) <= letter_counter[letter] <= int(rule_max):
            count = count + 1

    return count


def part2(rows):
    count = 0
    for line in rows:
        position1, position2, letter, password = parse_line(line)
        
        length = len(password)
        
        # offset the index by one since the authentication system is array starts at one
        index_a = int(position1) - 1
        index_b = int(position2) - 1

        is_first = index_a < length and password[index_a] == letter
        is_second = index_b < length and password[index_b] == letter

        # xor the results so that only one is valid
        if is_first ^ is_second:
            count = count + 1
    return count


if __name__ == "__main__":
    # Open the file and readlines
    with open('input') as file:
        rows = file.readlines()

        print("The first answer is: {}".format(part1(rows)))
        print("The second answer is: {}".format(part2(rows)))

File: rust/2/src/main.rs
use std::fs::File;
use std::io;
use std::io::{BufRead, BufReader};

fn read_input() -> Result<Vec<String>, io::Error> {
    let filename = "input";

    let file = File::open(filename)?;

    let reader = BufReader::new(file);
    let input: Vec<String> = reader.lines().collect::<Result<_, _>>().unwrap();

    Ok(input)
}

fn parse_line(line: &str) -> (usize, usize, char, &str) {
    let splitted_line: Vec<&str> = line.split([' ', ':', '-'].as_ref()).collect();
    let first_integer = splitted_line[0].parse::<usize>().unwrap();
    let second_integer = splitted_line[1].parse::<usize>().unwrap();
    let character = splitted_line[2].chars().nth(0).unwrap();
    let password = splitted_line[4];

    return (first_integer, second_integer, character, password);
}

fn part_1() -> i32 {
    let input = read_input();
    if let Err(e) = &input {
        eprintln!("Error occured: {}", e);
        std::process::exit(1);
    }
    
    let mut ok_count = 0;
    for line in input.unwrap().iter() {
        let (min, max, character, password) = parse_line(line);
        let count = password.chars().filter(|&c| c == character).count();
        
        if min <= count && count <= max {
            ok_count += 1;
        }
    }
    return ok_count;
}

fn part_2() -> i32 {
    let input = read_input();

    if let Err(e) = &input {
        eprintln!("Error occured: {}", e);
        std::process::exit(1);
    }

    let mut ok_count = 0;
    for line in input.unwrap().iter() {
        let (position1, position2, character, password) = parse_line(line);
        if (password.chars().nth(position1 - 1).unwrap() == character) 
            != (password.chars().nth(position2 - 1).unwrap() == character) {
                ok_count += 1;
            }
    }
    return ok_count;
}

fn main() {
    let part1_result = part_1();
    println!("The first answer is: {}", part1_result);
    let part2_result = part_2();
    println!("The second answer is: {}", part2_result);
}
